SIL.Archiving/RampArchivingDlgViewModel.cs

```diff
@@ -329,27 +292,24 @@ private void AddModesToSet(IEnumerable<string> files)
 		/// <param name="appName">The application name</param>
 		/// <param name="title">Title of the submission</param>
 		/// <param name="id">Identifier (used as filename) for the package being created</param>
-		/// <param name="appSpecificArchivalProcessInfo">Application can use this to pass
-		/// additional information that will be displayed to the user in the dialog to explain
-		/// any application-specific details about the archival process.</param>
 		/// <param name="setFilesToArchive">Delegate to request client to call methods to set
-		/// which files should be archived (this is deferred to allow display of progress message)</param>
+		/// which files should be archived (this is deferred to allow display of progress message).
+		/// Clients will normally do this by calling AddFileGroup one or more times.</param>
 		/// <param name="getFileDescription">Callback function to get a file description based
 		/// on the file-list key (param 1) and the filename (param 2)</param>
 		/// ------------------------------------------------------------------------------------
 		public RampArchivingDlgViewModel(string appName, string title, string id,
-			string appSpecificArchivalProcessInfo, Action<ArchivingDlgViewModel> setFilesToArchive,
+			Action<ArchivingDlgViewModel, CancellationToken> setFilesToArchive,
 			Func<string, string, string> getFileDescription) : base(appName, title, id,
-			appSpecificArchivalProcessInfo, setFilesToArchive)
+			(m, c) => { ((RampArchivingDlgViewModel)m).SetFilesToArchive(c); })
 		{
-			if (getFileDescription == null)
-				throw new ArgumentNullException("getFileDescription");
-			_getFileDescription = getFileDescription;
+			_getFileDescription = getFileDescription ?? throw new ArgumentNullException(nameof(getFileDescription));
+			_setFilesToArchive = setFilesToArchive;
```

Comment:
There might be a better way to do this that would be simpler, but the derived class passes its method (which does more than just call the passed in action) as a delegate into the base class, which stores that in its version of the field. which it turns into an asynchronous call, so as it stands, it would not be sufficient to merely override the member in the derived class.

Reply:
Here's what I'm thinking. In ArchivingDlgViewModel, make `Task SetFilesToArchive` protected and virtual. Then in this Ramp class we override it like this:
```C#
protected override async Task SetFilesToArchive(CancellationToken cancellationToken)
{
	await base.SetFilesToArchive(cancellationToken);
	if (cancellationToken.IsCancellationRequested)
		throw new OperationCanceledException();
	foreach (var fileList in FileLists.Where(fileList => fileList.Value.Item1.Any()))
	{
		var normalizedName = NormalizeFilename(fileList.Key,
			Path.GetFileName(fileList.Value.Item1.First()));
		_progressMessages[normalizedName] = fileList.Value.Item2;
	}
}
```

calling base.SetFilesToArchive will still call the `setFilesToArchive` delegate passed in to the constructor, and because `SetFilesToArchive` was overriden when `ArchivingDlgViewModel.Initialize` calls `SetFilesToArchive` it will actually call the on defined by the Ramp class. Unless I'm missing something this does the same thing.

Reply:
Makes sense. Done!